dask/array/core.py

```diff
@@ -2062,7 +2062,7 @@ def asarray(array):
     """
     if not isinstance(array, Array):
         name = 'asarray-' + tokenize(array)
-        if not hasattr(array, 'shape'):
+        if isinstance(getattr(array, 'shape', None), Iterable):
```

Comment:
I think this does the wrong thing now. Previously, `asarray` would not convert an HDF5 dataset (for example) into a NumPy array but now it does.